[chainerx] add `prod`

kshitij12345 opened the pull request:
This PR intends to add Prod to ChainerX.

It is implemented with reference of Sum.

Done :
* Native Device Implementation.
* Cuda Device Implementation.
* Backward Function.
* Python bindings for `chx.prod`.
* Python bindings for `array.prod`.
* Tests

**TO-DO**
~~Update Documentation~~

kshitij12345 commented:
8 tests for compute_prod fail. All are the cases where `dtype` is `float16` and `device` is `cuda`.
![image](https://user-images.githubusercontent.com/19503980/55189283-81394f00-51c3-11e9-94fd-68c489bce311.png)


Numpy returns `0.0` while ChainerX returns `nan`.
The inputs for the failed cases are [here](https://github.com/chainer/chainer/blob/fb35b515ef05d58c79a1c390e1b7f716a199f71e/tests/chainerx_tests/unit_tests/routines_tests/test_math.py#L1277-L1278).

From my guess, the hotspots should maybe [here](https://github.com/chainer/chainer/blob/fb35b515ef05d58c79a1c390e1b7f716a199f71e/chainerx_cc/chainerx/cuda/cuda_device/reduction.cu#L91) or [here](https://github.com/chainer/chainer/blob/c5b075ba46c285c0bcf7466d77c6abda5c158617/chainerx_cc/chainerx/cuda/float16.cuh#L55)

Note : All other types for the same inputs pass.

Please Help.

kshitij12345 commented:
@niboshi , could you help me figure why double_backward is failing but not backward? I am not sure which part of backward is wrong exactly. Thank You.

niboshi commented on chainerx_cc/chainerx/routines/math.cc:
```diff
@@ -477,6 +477,50 @@ Array Sum(const Array& a, const OptionalAxes& axis, bool keepdims) {
+
+    BackwardBuilder bb{"prod", a, out};
+    if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+        bt.Define([sorted_axis, a, out, in_shape = a.shape(), keepdims](BackwardContext& bctx) {
```
Not sure if it's relevant, but I think `a` and `out` should be captured using `bb.RetainInput(0)` and `bb.RetainOutput(0)` respectively?

kshitij12345 replied:
Done.

niboshi commented on chainerx_cc/chainerx/native/native_device/reduction.cc:
```diff
@@ -76,6 +76,31 @@ class NativeSumKernel : public SumKernel {
+            struct Impl {
+                Accum Identity() { return Accum{1}; }
+                Accum MapIn(In in, int64_t /*index*/) { return static_cast<Accum>(in); }
+                void Reduce(Accum next, Accum& accum) { accum *= next; }
```
(This is not relevant to the error)
It won't compile in gcc, complaining `*=` is invalid for `bool`s.

As `Accum` (and `Out`) is never actually `bool`, let's replace `VisitDtype` (the first one) at the line 98 with `VisitNumericDtype`, so that compilers won't try to generate the code for `bool`.

https://github.com/chainer/chainer/pull/6687/files#diff-5cc617697b5522640bcf9858d0c42b73R98

kshitij12345 replied:
Done.

kshitij12345 commented:
@niboshi , after merging the latest master and  addressing the above comments, the `double_backward` issue is resolved except for the  few random `float16` cases.

The `double_backward` for `float16` fails spurriously many times. Please review.

toslunar commented on chainerx_cc/chainerx/routines/math.cc:
```diff
@@ -513,6 +513,53 @@ Array Sum(const Array& a, const OptionalAxes& axis, bool keepdims) {
+                }
+                bctx.input_grad() = (out.Reshape(out_shape_broadcastable) * gout.Reshape(out_shape_broadcastable) / a);
+            } else {
+                bctx.input_grad() = out * gout / a;
```
Could you avoid `0 / 0`?  There are at least two ways (#3031, #5094) to implement backward that allows forward inputs to contain `0`.  It might be fine to leave `backward` of `chainerx.prod` totally unimplemented.

kshitij12345 replied:
Sure. Do note that it'll take me some time to resume work on this PR.

niboshi commented:
We're no longer merging feature PRs.
Let me close the PR.
